include/deal.II/lac/petsc_precondition.h

```diff
@@ -957,6 +957,131 @@ namespace PETScWrappers
     AdditionalData additional_data;
   };
 
+  /**
+   * A class that implements the interface to use the BDDC preconditioner from
+   * PETSc (<a
+   * href="https://petsc.org/release/docs/manualpages/PC/PCBDDC.html">PCBDDC</a>),
+   * which is a two-level, substructuring, non-overlapping domain decomposition
+   * preconditioner. Details of the implementation can be found in "S Zampini,
+   * SISC (2016)". It mainly consists of two elements:
+   *
+   * <ul>
+   *   <li> Local solvers: Solvers for each subdomain. These are performed concurrently by each processor
+   *   <li> A coarse solver: Continuity between each subdomain is imposed in a small number of DoFs, referred to as <em>primal DoFs</em>. This solver solves such problem.
+   * </ul>
+   *
+   * The size of the primal space is determined through the @p AdditionalData parameters.
+   *
+   * @ingroup PETScWrappers
+   */
+  class PreconditionBDDC : public PreconditionBase
+  {
+  public:
+    /**
+     * Standardized data struct to pipe additional flags to the
+     * preconditioner.
+     */
+    struct AdditionalData
+    {
+      /**
+       * Constructor. Note that BDDC offers a lot more options to set
+       * than what is exposed here.
+       */
+      AdditionalData(const bool                    use_vertices = true,
+                     const bool                    use_edges    = false,
+                     const bool                    use_faces    = false,
+                     const bool                    symmetric    = false,
+                     const unsigned int            coords_cdim  = 0,
+                     const types::global_dof_index coords_n     = 0,
+                     const PetscReal *             coords_data  = nullptr);
+      /**
+       * This flag sets the use of degrees of freedom in the vertices of the
+       * subdomains as primal variables for the creation of the coarse space.
+       */
+      bool use_vertices;
+
+      /**
+       * This flag sets the use of degrees of freedom in the edges of the
+       * subdomain as primal variables for the creation of the coarse space.
+       * Continuity is actually imposed at the edge average.
+       */
+      bool use_edges;
+
+      /**
+       * This flag sets the use of degrees of freedom in the faces of the
+       * subdomain as primal variables for the creation of the coarse space.
+       * Continuity is actually imposed at the face average.
+       */
+      bool use_faces;
+
+      /**
+       * Set whether the matrix is symmetric or not.
+       */
+      bool symmetric;
+
+      /**
+       * Set the number of coordinates that each DoF has, i.e. 2 in 2D and 3 in
+       * 3D.
+       */
+      unsigned int coords_cdim;
```

Comment:
Same thing here - this should just be a compile-time constant (spacedim).

Reply:
Included in previous mod